pkg/pipeline/encode/encode_prom.go

```diff
@@ -196,44 +235,48 @@ func (e *EncodeProm) cleanupExpiredEntriesLoop() {
 	}
 }
 
-// startPrometheusInterface listens for prometheus resource usage requests
-func startPrometheusInterface(w *EncodeProm) {
-	log.Debugf("entering startPrometheusInterface")
-	log.Infof("startPrometheusInterface: port num = %s", w.port)
+// startServer listens for prometheus resource usage requests
+func (e *EncodeProm) startServer() {
+	log.Debugf("entering startServer")
 
 	// The Handler function provides a default handler to expose metrics
 	// via an HTTP server. "/metrics" is the usual endpoint for that.
 	http.Handle("/metrics", promhttp.Handler())
 
-	err := http.ListenAndServe(w.port, nil)
-	if err != nil {
+	err := e.server.ListenAndServe()
+	if err != nil && err != http.ErrServerClosed {
 		log.Errorf("error in http.ListenAndServe: %v", err)
 		os.Exit(1)
 	}
 }
 
+func (e *EncodeProm) closeServer(ctx context.Context) error {
+	return e.server.Shutdown(ctx)
+}
+
 func NewEncodeProm(params config.StageParam) (Encoder, error) {
-	jsonEncodeProm := api.PromEncode{}
+	config := api.PromEncode{}
 	if params.Encode != nil && params.Encode.Prom != nil {
-		jsonEncodeProm = *params.Encode.Prom
+		config = *params.Encode.Prom
 	}
 
-	portNum := jsonEncodeProm.Port
-	promPrefix := jsonEncodeProm.Prefix
-	expiryTime := int64(jsonEncodeProm.ExpiryTime)
+	expiryTime := int64(config.ExpiryTime)
 	if expiryTime == 0 {
 		expiryTime = defaultExpiryTime
 	}
 	log.Debugf("expiryTime = %d", expiryTime)
 
-	metrics := make(map[string]metricInfo)
-	for _, mInfo := range jsonEncodeProm.Metrics {
-		var pMetric PromMetric
-		fullMetricName := promPrefix + mInfo.Name
+	counters := []counterInfo{}
+	gauges := []gaugeInfo{}
+	histos := []histoInfo{}
+	aggHistos := []histoInfo{}
+
+	for i := range config.Metrics {
+		mInfo := config.Metrics[i]
```

Comment:
yes, I don't know why I wrote it that way

Reply:
Now I remember: that's because a pointer to `mInfo` is use below, so we need to do a copy somewhere. But I'll do differently, I'll change the model to not use a pointer, so the copy is implicit